In [26]:

import pandas as pd 
import numpy as np
import re

## ETL Extracción, transformación y carga

- HOCIMIDIOS 

In [3]:
homicidios_hechos = pd.read_excel("Datasets/raw/homicidios.xlsx", sheet_name="HECHOS")
homicidios_victimas = pd.read_excel("Datasets/raw/homicidios.xlsx", sheet_name="VICTIMAS")

In [22]:
homicidios_hechos.head(3)

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00   4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00   7   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE             Calle  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  PIEDRA BUENA AV.   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ    PAZ, GRAL. AV.   
2                         AV ENTRE RIOS 2034       AVENIDA    ENTRE RIOS AV.   

                                 Cruce  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                  DE LOS CORRALES AV.   
2                                  NaN   

                               Dirección Normalizada  COMUNA  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...       8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.       9   
2                                ENTRE RIOS AV. 2034       1   

                                XY (CABA)      pos x      pos y  \
0   Point (98896.78238426 93532.43437792) -58.475340 -34.687570   
1   Point (95832.05571093 95505.41641999) -58.508775 -34.669777   
2  Point (106684.29090040 99706.57687843) -58.390403 -34.631894   

    PARTICIPANTES VICTIMA    ACUSADO  
0       MOTO-AUTO    MOTO       AUTO  
1  AUTO-PASAJEROS    AUTO  PASAJEROS  
2       MOTO-AUTO    MOTO       AUTO

In [6]:
homicidios_hechos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     696 non-null    object        
 1   N_VICTIMAS             696 non-null    int64         
 2   FECHA                  696 non-null    datetime64[ns]
 3   AAAA                   696 non-null    int64         
 4   MM                     696 non-null    int64         
 5   DD                     696 non-null    int64         
 6   HORA                   696 non-null    object        
 7   HH                     696 non-null    object        
 8   LUGAR_DEL_HECHO        696 non-null    object        
 9   TIPO_DE_CALLE          696 non-null    object        
 10  Calle                  695 non-null    object        
 11  Altura                 129 non-null    float64       
 12  Cruce                  525 non-null    object        
 13  Direc

Examen de nulos y duplicados

In [7]:
homicidios_hechos.isnull().sum()

ID                         0
N_VICTIMAS                 0
FECHA                      0
AAAA                       0
MM                         0
DD                         0
HORA                       0
HH                         0
LUGAR_DEL_HECHO            0
TIPO_DE_CALLE              0
Calle                      1
Altura                   567
Cruce                    171
Dirección Normalizada      8
COMUNA                     0
XY (CABA)                  0
pos x                      0
pos y                      0
PARTICIPANTES              0
VICTIMA                    0
ACUSADO                    0
dtype: int64

- Se analizan las columnas con datos nulos

In [10]:
homicidios_hechos[["Altura", "Calle", "Dirección Normalizada", "Cruce"]].head(5)

Altura             Calle  \
0     NaN  PIEDRA BUENA AV.   
1     NaN    PAZ, GRAL. AV.   
2  2034.0    ENTRE RIOS AV.   
3     NaN    LARRAZABAL AV.   
4     NaN      SAN JUAN AV.   

                               Dirección Normalizada  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.   
2                                ENTRE RIOS AV. 2034   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.   

                                 Cruce  
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.  
1                  DE LOS CORRALES AV.  
2                                  NaN  
3             VILLEGAS, CONRADO, GRAL.  
4              SAENZ PE?A, LUIS, PRES.

In [11]:
homicidios_hechos[homicidios_hechos["Dirección Normalizada"].isnull()][["Dirección Normalizada", "Cruce","Altura", "Calle"]]

Dirección Normalizada Cruce  Altura                                 Calle
38                    NaN   NaN     NaN                 LUGONES, LEOPOLDO AV.
106                   NaN   NaN     NaN     AUTOPISTA BUENOS AIRES - LA PLATA
119                   NaN   NaN     NaN                                   NaN
180                   NaN   NaN     NaN               AUTOPISTA PERITO MORENO
181                   NaN   NaN     NaN  AUTOPISTA DELLEPIANE LUIS TTE. GRAL.
313                   NaN   NaN     NaN                 LUGONES, LEOPOLDO AV.
546                   NaN   NaN     NaN                 LUGONES, LEOPOLDO AV.
621                   NaN   NaN     NaN     AUTOPISTA BUENOS AIRES - LA PLATA

Tras examinar datos nulos, que representan el 5,1% del dataset:
- **`Calle`** tiene un solo dato nulo.
- **`Dirección Normalizada`** 8 nulos. 
- **`Cruce`** tiene 171 nulos.
- **`Altura`** tiene 567 nulos.

Analizando los datos de la `Altura` se procede a eliminarla debido a la gran cantidad de faltantes y no existiendo manera de poder rellenar las colummas; ya que altura se refiere a la numeración de la calle donde se produce el siniestro y la mayoria de los casos se producen en cruces o esquinas.
La columna `Cruce` solo va a tener dato cuando el hecho se porduce en un cruce de calles, por lo que se conserva de esta manera.
La columna `Dirección Normalizada` no tiene posibilidad de completarse con las columnas del dataset; pero contiene solo 1,1% de nulos por lo se conserva de esta manera.

Se elimina la columna "Altura"

In [12]:
homicidios_hechos= homicidios_hechos.drop("Altura", axis=1)
homicidios_hechos.columns

Index(['ID', 'N_VICTIMAS', 'FECHA', 'AAAA', 'MM', 'DD', 'HORA', 'HH',
       'LUGAR_DEL_HECHO', 'TIPO_DE_CALLE', 'Calle', 'Cruce',
       'Dirección Normalizada', 'COMUNA', 'XY (CABA)', 'pos x', 'pos y',
       'PARTICIPANTES', 'VICTIMA', 'ACUSADO'],
      dtype='object')

Duplicados

In [13]:
homicidios_hechos.duplicated().sum()

0

Se transforman los datos de la columna 'HH' a tipo numérico.

In [15]:
homicidios_hechos["HH"].unique()

array([4, 1, 7, 0, 5, 18, 19, 15, 11, 22, 16, 9, 23, 6, 10, 17, 12, 8, 20,
       21, 14, 3, 2, 13, 'SD'], dtype=object)

In [16]:
# Cambio el dato horas  a tipo de dato numerico y luego a entero 
homicidios_hechos["HH"] = pd.to_numeric(homicidios_hechos ["HH"], errors="coerce")
homicidios_hechos["HH"] = homicidios_hechos["HH"].astype("Int64")

In [17]:
homicidios_hechos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     696 non-null    object        
 1   N_VICTIMAS             696 non-null    int64         
 2   FECHA                  696 non-null    datetime64[ns]
 3   AAAA                   696 non-null    int64         
 4   MM                     696 non-null    int64         
 5   DD                     696 non-null    int64         
 6   HORA                   696 non-null    object        
 7   HH                     695 non-null    Int64         
 8   LUGAR_DEL_HECHO        696 non-null    object        
 9   TIPO_DE_CALLE          696 non-null    object        
 10  Calle                  695 non-null    object        
 11  Cruce                  525 non-null    object        
 12  Dirección Normalizada  688 non-null    object        
 13  COMUN

Transformacion a columnas 'pos x' y 'pos y'

In [20]:
# Filtrar filas con posiciones no válidas
valid_positions = ~(homicidios_hechos['pos x'].str.contains('[^0-9.,-]') | homicidios_hechos['pos y'].str.contains('[^0-9.,-]') | (homicidios_hechos['pos x'] == '.') | (homicidios_hechos['pos y'] == '.'))

# Establecer posiciones no válidas en NaN
homicidios_hechos.loc[~valid_positions, ['pos x', 'pos y']] = np.nan

# Convertir las cadenas a tipo numérico (float)
homicidios_hechos['pos x'] = homicidios_hechos['pos x'].str.replace(',', '.').astype(float)
homicidios_hechos['pos y'] = homicidios_hechos['pos y'].str.replace(',', '.').astype(float)

In [21]:
homicidios_hechos.head(3)

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00   4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00   7   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE             Calle  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  PIEDRA BUENA AV.   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ    PAZ, GRAL. AV.   
2                         AV ENTRE RIOS 2034       AVENIDA    ENTRE RIOS AV.   

                                 Cruce  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                  DE LOS CORRALES AV.   
2                                  NaN   

                               Dirección Normalizada  COMUNA  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...       8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.       9   
2                                ENTRE RIOS AV. 2034       1   

                                XY (CABA)      pos x      pos y  \
0   Point (98896.78238426 93532.43437792) -58.475340 -34.687570   
1   Point (95832.05571093 95505.41641999) -58.508775 -34.669777   
2  Point (106684.29090040 99706.57687843) -58.390403 -34.631894   

    PARTICIPANTES VICTIMA    ACUSADO  
0       MOTO-AUTO    MOTO       AUTO  
1  AUTO-PASAJEROS    AUTO  PASAJEROS  
2       MOTO-AUTO    MOTO       AUTO

In [24]:
homicidios_hechos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     696 non-null    object        
 1   N_VICTIMAS             696 non-null    int64         
 2   FECHA                  696 non-null    datetime64[ns]
 3   AAAA                   696 non-null    int64         
 4   MM                     696 non-null    int64         
 5   DD                     696 non-null    int64         
 6   HORA                   696 non-null    object        
 7   HH                     695 non-null    Int64         
 8   LUGAR_DEL_HECHO        696 non-null    object        
 9   TIPO_DE_CALLE          696 non-null    object        
 10  Calle                  695 non-null    object        
 11  Cruce                  525 non-null    object        
 12  Dirección Normalizada  688 non-null    object        
 13  COMUN

In [27]:
# Función para extraer las coordenadas CABA
# Esta función toma un argumento texto, que es el texto de entrada del cual se intentarán extraer las coordenadas
# Utiliza una expresión regular (re.findall) para buscar todas las coincidencias de números decimales (que tienen un punto como separador decimal) en el texto. Estas coincidencias se almacenan en la lista coordenadas.
# return float(coordenadas[0]), float(coordenadas[1]): Si se encontraron dos coordenadas, se convierten en números de punto flotante (float) y se devuelven como una tupla de dos valores.
# Luego, se aplica esta función a la columna 'XY (CABA)' de df_siniestros utilizando el método .map(). Esto se hace para cada fila de la columna, y los resultados se almacenan en dos nuevas columnas, 'coordenada_x' y 'coordenada_y', utilizando zip(*...).

def extraer_coordenadas(texto):
    # Utilizamos una expresión regular para extraer las coordenadas
    coordenadas = re.findall(r'\d+\.\d+', texto)
    if len(coordenadas) == 2:
        return float(coordenadas[0]), float(coordenadas[1])
    else:
        return None, None

# Aplicar la función a la columna 'XY (CABA)'
homicidios_hechos['coordenada_x'], homicidios_hechos['coordenada_y'] = zip(*homicidios_hechos['XY (CABA)'].map(extraer_coordenadas))

# Filtrar los registros que no tienen coordenadas válidas
homicidios_hechos = homicidios_hechos[(homicidios_hechos['coordenada_x'].notnull()) & (homicidios_hechos['coordenada_y'].notnull())]

In [29]:
homicidios_hechos.head(2)

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00   4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...   

                               Dirección Normalizada COMUNA  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   

                               XY (CABA)      pos x      pos y  \
0  Point (98896.78238426 93532.43437792) -58.475340 -34.687570   
1  Point (95832.05571093 95505.41641999) -58.508775 -34.669777   

    PARTICIPANTES  VICTIMA    ACUSADO  coordenada_x  coordenada_y  
0       MOTO-AUTO     MOTO       AUTO  98896.782384  93532.434378  
1  AUTO-PASAJEROS     AUTO  PASAJEROS  95832.055711  95505.416420  

[2 rows x 22 columns]

Columnas resultantes

In [34]:
homicidios_hechos.columns

Index(['ID', 'N_VICTIMAS', 'FECHA', 'AAAA', 'MM', 'DD', 'HORA', 'HH',
       'LUGAR_DEL_HECHO', 'TIPO_DE_CALLE', 'Calle', 'Cruce',
       'Dirección Normalizada', 'COMUNA', 'XY (CABA)', 'pos x', 'pos y',
       'PARTICIPANTES', 'VICTIMA', 'ACUSADO', 'coordenada_x', 'coordenada_y'],
      dtype='object')

Creo columna Semestre

In [35]:
# Definir una función para determinar el número de semestre
def obtener_semestre(row):
    return 1 if row['FECHA'].month <= 6 else 2

# Aplicar la función a la columna 'FECHA' para obtener la columna 'SEMESTRE'
homicidios_hechos['SEMESTRE'] = homicidios_hechos.apply(obtener_semestre, axis=1)


In [36]:
homicidios_hechos

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
0    2016-0001           1 2016-01-01  2016   1   1  04:00:00   4   
1    2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   
2    2016-0003           1 2016-01-03  2016   1   3  07:00:00   7   
3    2016-0004           1 2016-01-10  2016   1  10  00:00:00   0   
4    2016-0005           1 2016-01-21  2016   1  21  05:20:00   5   
..         ...         ...        ...   ...  ..  ..       ...  ..   
691  2021-0093           1 2021-12-13  2021  12  13  17:10:00  17   
692  2021-0094           1 2021-12-20  2021  12  20  01:10:00   1   
693  2021-0095           1 2021-12-30  2021  12  30  00:43:00   0   
694  2021-0096           1 2021-12-15  2021  12  15  10:30:00  10   
695  2021-0097           1 2021-11-18  2021  11  18  06:10:00   6   

                               LUGAR_DEL_HECHO TIPO_DE_CALLE  ... COMUNA  \
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...      8   
1            AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...      9   
2                           AV ENTRE RIOS 2034       AVENIDA  ...      1   
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...      8   
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...      1   
..                                         ...           ...  ...    ...   
691                          AV. RIESTRA Y MOM       AVENIDA  ...      7   
692                    AU DELLEPIANE Y LACARRA     AUTOPISTA  ...      9   
693                        AV. GAONA Y TERRADA       AVENIDA  ...     11   
694                         AV. EVA PERON 4071       AVENIDA  ...      9   
695                    PADRE CARLOS MUJICA 709         CALLE  ...      1   

                                   XY (CABA)      pos x      pos y  \
0      Point (98896.78238426 93532.43437792) -58.475340 -34.687570   
1      Point (95832.05571093 95505.41641999) -58.508775 -34.669777   
2     Point (106684.29090040 99706.57687843) -58.390403 -34.631894   
3      Point (99840.65224780 94269.16534422) -58.465039 -34.680930   
4    Point (106980.32827929 100752.16915795) -58.387183 -34.622466   
..                                       ...        ...        ...   
691   Point (102728.60090138 98186.24929177) -58.433538 -34.645616   
692    Point (99624.29795829 97569.69801131) -58.467398 -34.651178   
693   Point (99116.45492358 101045.23284826) -58.472934 -34.619847   
694    Point (99324.54463985 97676.26932409) -58.470668 -34.650217   
695  Point (107664.16647795 104708.63962087) -58.379762 -34.586796   

      PARTICIPANTES    VICTIMA    ACUSADO   coordenada_x   coordenada_y  \
0         MOTO-AUTO       MOTO       AUTO   98896.782384   93532.434378   
1    AUTO-PASAJEROS       AUTO  PASAJEROS   95832.055711   95505.416420   
2         MOTO-AUTO       MOTO       AUTO  106684.290900   99706.576878   
3           MOTO-SD       MOTO         SD   99840.652248   94269.165344   
4    MOTO-PASAJEROS       MOTO  PASAJEROS  106980.328279  100752.169158   
..              ...        ...        ...            ...            ...   
691       MOTO-AUTO       MOTO       AUTO  102728.600901   98186.249292   
692       MOTO-AUTO       MOTO       AUTO   99624.297958   97569.698011   
693     MOTO-CARGAS       MOTO     CARGAS   99116.454924  101045.232848   
694     AUTO-CARGAS       AUTO     CARGAS   99324.544640   97676.269324   
695  BICICLETA-AUTO  BICICLETA       AUTO  107664.166478  104708.639621   

    SEMESTRE  
0          1  
1          1  
2          1  
3          1  
4          1  
..       ...  
691        2  
692        2  
693        2  
694        2  
695        2  

[682 rows x 23 columns]